[OpenMP][MLIR] Add num_teams clause with dims modifier support

skc7 opened the pull request:
PR adds support of openmp 6.1 feature `num_teams` with dims modifier.
llvmIR translation for num_teams with dims modifier is marked as NYI.

skc7 committed:
[OpenMP][MLIR] Add num_teams clause with dims modifier support

skc7 committed:
fix comment

mjklemm commented on mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td:
```diff
@@ -1532,4 +1532,76 @@ class OpenMP_UseDevicePtrClauseSkip<
 def OpenMP_UseDevicePtrClause : OpenMP_UseDevicePtrClauseSkip<>;
 
+//===----------------------------------------------------------------------===//
+// V6.2: Multidimensional `num_teams` clause with dims modifier
```
```suggestion
// V6.1: Multidimensional `num_teams` clause with dims modifier
```

mjklemm commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -4453,6 +4454,68 @@ LogicalResult WorkdistributeOp::verify() {
+  // If dims not specified but we have values, it's implicitly unidimensional
+  if (!dims.has_value() && values.size() != 1) {
+    return parser.emitError(parser.getCurrentLocation())
+           << "expected 1 value without dims modifier, got " << values.size();
```
```suggestion
           << "expected 1 value without dims modifier, but got " << values.size() << " values";
```

skc7 replied:
Updated. Thanks

skc7 committed:
Comments fix

skc7 commented on mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td:
```diff
@@ -1532,4 +1532,76 @@ class OpenMP_UseDevicePtrClauseSkip<
+  }];
+}
+
+def OpenMP_NumTeamsMultiDimClause : OpenMP_NumTeamsMultiDimClauseSkip<>;
```
Original design was to have a separate dims modifier(with dims and values args) class and then create `num_teams` and `thread_limit` clauses from it. But this leads to both clauses having the same argument names and when added to `teams` Op would create an issue. 

So, now created just `num_teams_multi_dim` clause with arguments as `num_teams_dims` and `num_teams_values`. 
Will remove the old `num_teams` clause and replace it with `num_teams_multi_dim` clause and move the name back to `num_teams`

skc7 committed:
Use DimsModifier for custom assembly parser and printer

skc7 committed:
use dims modifer in main num_teams clause itself instead of creating new clause

skc7 commented:
Updated PR to use dims modifier arguments in the original `num_teams` clause itself.
This is to support the old version and new version of `num_teams` using the same clause itself.
- omp.teams num_teams(%lb : i32 to %ub : i32)
- omp.teams num_teams(dims(3): %lb, %ub, %ub : i32)

Updated parser, printer and verifier for the clause.

Thanks for feedback @kparzysz

skc7 committed:
Mark mlir->llvmir for num_teams with dims as NYI

skc7 committed:
few more fixes

skc7 committed:
allow single dims_values

skc7 committed:
Make dims_values as IntLikeType

mjklemm commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -6035,6 +6040,10 @@ extractHostEvalClauses(omp::TargetOp targetOp, Value &numThreads,
+            // num_teams dims and values are not yet supported
+            assert(!teamsOp.hasNumTeamsDimsModifier() &&
+                   "Lowering of num_teams with dims modifier is not yet "
+                   "implemented.");
```
This eventually caught my eye.  I have seen other PRs that make use `TODO(loc,...)` instead of an `assert`.  I have started to wonder if we should do the same in this stream of PRs (also in the other ones that I have approved already)

skatrak replied:
Yes, not-yet-implemented errors in this pass should be detected and notified to users by `checkImplementationStatus()`, which then causes it to stop early and exit cleanly.

skc7 replied:
have added `checkNumTeamsMultiDim` to check the status of lowering instead of assert. Updated it in latest patch for all the clauses.

mjklemm reviewed (approved):
Other than the stylistic comment about `assert`, the PR LGTM.  I will let @kparzysz comment on the stylistic question and maybe request a change.

skc7 committed:
Update num_teams to have just the list and no dims(N) syntax

kevinsala commented on mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td:
```diff
@@ -992,38 +991,39 @@ class OpenMP_NumTeamsClauseSkip<
 
     /// Returns the number of dimensions specified for num_teams
     unsigned getNumTeamsDimsCount() {
-      if (!hasNumTeamsDimsModifier())
```
Why is the `num_teams` not having the `dims` modifier anymore?

skc7 replied:
`num_teams`, `num_threads`, `thread_limit` are now updated to take a list of integers for dims values.

Flang frontend would have the syntax `num_teams(dims(N) : 4, 8...)`, after parsing it and verifies for N to be equal to values specified, it is lowered to `num_teams` clause in MLIR which will only store the values, N is implicit and equals to size of list. So have removed the `dims(N)` syntax for MLIR clauses.

This is as per review comment recieved for https://github.com/llvm/llvm-project/pull/171825
LINK: https://github.com/llvm/llvm-project/pull/171825#discussion_r2690609602 
CC: @skatrak

kevinsala replied:
Got it! Thank you for the clarification @skc7!

skc7 committed:
name change to getNumTeams()

skc7 committed:
rename num_teams_vals to num_teams_vars

skatrak commented on mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td:
```diff
@@ -974,22 +974,57 @@ class OpenMP_NumTeamsClauseSkip<
   let arguments = (ins
+    Variadic<IntLikeType>:$num_teams_vars,
     Optional<AnyInteger>:$num_teams_lower,
     Optional<AnyInteger>:$num_teams_upper
```
I think that here we wouldn't want to have two different arguments to represent the upper bounds when in unidimensional and multidimensional cases.
```suggestion
    Optional<AnyInteger>:$num_teams_lower,
    Variadic<AnyInteger>:$num_teams_upper_vars
```
If my understanding of the spec is correct, the cases that are allowed are:
- Unidimensional: There's exactly one `num_teams_upper_vars` element and, optionally, `num_teams_lower` might be specified.
- Multidimensional: There's more than one element in `num_teams_upper_vars` and `num_teams_lower` mustn't be specified.

skatrak commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -2634,6 +2635,35 @@ void TeamsOp::build(OpBuilder &builder, OperationState &state,
+    return op->emitError()
+           << "num_teams multi-dimensional values cannot be used together with "
+              "legacy lower/upper bounds";
+  }
```
No need to check this if we don't have a redundant representation.
```suggestion
```

skatrak reviewed (commented):
Thank you Chainanya, I've got a comment about the chosen representation. I think we can hopefully make this patch simpler as a result.

skc7 committed:
Use only num_teams_lower and num_teams_upper_vars for clause

skc7 commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -2634,6 +2635,35 @@ void TeamsOp::build(OpBuilder &builder, OperationState &state,
+  }
+
+  // If lower is specified, upper must also be specified
+  if (numTeamsLower) {
```
Added the check for numTeamsUpperVars.size() != 1

skatrak reviewed (commented):
Thank you again, this LGTM. I have just one comment about the pre-MLIR representation of the clause.

skc7 committed:
Remove space in test

skc7 committed:
Update NumTeamsT clause to accept upper bound list

skatrak reviewed (approved):
LGTM, thank you. Please wait for @kparzysz to take a look, in case there are any concerns about the changes to `NumTeamsT` or if a new semantic check for this should be added.
